### ISSUE TITLE
Secrets being deployed after Deployments

### ISSUE THREAD

**Issue by colemickens** [pulumi/pulumi-kubernetes#79]
I have a fairly straightforward module I'm trying to write for Dex.

It's currently waiting on the Deployment to succeed before proceeding, but the Deployment depends on the Secret.

I don't see an explicit way to link them, going off of the MariaDB example... https://github.com/pulumi/pulumi-kubernetes/blob/master/examples/mariadb/index.ts It seems even in this example, that it's a loose associate via k8s at runtime. I don't see anything that would let pulumi know there's a relationship.

If I reference the Secret from the Deployment object somehow, will it work? Is there an example of this?

**Comment by colemickens** [comment]
(Note, I don't really mind if the Secret gets created first or second, but the Deployment is just waiting on something that will never succeed, as-is)

**Comment by colemickens** [comment]
Looks related to #65. @hausdorff, you call out that "declaring dependencies between kube objects is kind of annoying" but the only example you given is the one that doesn't work since the change was backed off.

Is there a workaround in the meantime?

**Comment by hausdorff** [comment]
The easiest thing to do right now looks something like:

```
const secret = new k8s.core.v1.Secret("my-secret", secretJson);
const depl = new k8s.apps.v1.Deployment("my-depl", deplJson, {dependsOn: secret});
```

Eventually we will allow for direct references. For simplicity, you can imagine a fake `KubeObject` kind like this:

```
const secret = new k8s.core.v1.Secret("my-secret", secretJson);
const depl = new k8s.some.v1.KubeObject("my-obj", {useThisSecret: secret.metadata.apply(m => m.name)});
```

Pulumi will then "magically" figure out the dependency.

The reason this isn't done now is because the k8s API is super dense, and we wanted to be deliberate about making sure it was awesome.

**Comment by colemickens** [comment]
Aha! Thank you!

I was just now trying:

```
volumes: [{
  name: "config",
  secret: { secretName: dexSecret.metadata.apply<string>(m => m.name) },
}],  
```

but was getting: `Type 'Output<string>' is not assignable to type 'string'.` Sounds like it's in-progress / under consideration.

I'll try your suggested workaround for now, doesn't seem that bad to me anyway.

**Comment by colemickens** [comment]
I added `{ dependsOn: [dexSecret]}`, but it doesn't do what I expect, the Deployment is still going first.

Source: https://gist.github.com/colemickens/7127d334cb105f2062850a58e67166dd#file-index-ts-L70

**Comment by colemickens** [comment]
Actually, no, it worked as described. (But Pulumi thought the Deployment existed when I'd deleted it out from under it. I probably need to adjust how I work and cleanup... I'm accustomed to always blowing everything away and redeploying at once, rather than incrementally in the face of errors)

**Comment by hausdorff** [comment]
Yeah, do also note that incremental updates are not great right now for the Kubernetes provider. (_cf_., [my comment on #992](https://github.com/pulumi/pulumi/issues/992#issuecomment-399611875)). Basically, Pulumi does not have a notion of a resource that was successfully _created_ but which failed to _initialize_. The good news is that I do have an almost-ready patch to mostly-fix this, though -- I consider this to be the worst known issue with the model as it relates to Kubernetes, so it's a top priority for me to fix.

**Comment by hausdorff** [comment]
Since I think the issue is mostly a duplicate of #65, I'll close this for now. Feel free to re-open if you think this is wrong.
